bazel/PPROF.md

```diff
@@ -155,3 +175,145 @@ More complex flame/graph charts can be generated and viewed in a browser, which
+        .file   ""
+        .text
+        .globl  _Znwm
+        .type   _Znwm, @function
+_Znwm:
+.LFB73:
+        .cfi_startproc
+        pushq   %rbp
+        .cfi_def_cfa_offset 16
+        .cfi_offset 6, -16
+        movq    %rsp, %rbp
+        .cfi_def_cfa_register 6
+        movq    %rdi, -8(%rbp)
+        nop
+        popq    %rbp
+        .cfi_def_cfa 7, 8
+        ret
+        .cfi_endproc
+.LFE73:
+        .size   _Znwm, .-_Znwm
+        .ident  "GCC: (GNU) 10.2.1 20201016 (Red Hat 10.2.1-6)"
+        .section        .note.GNU-stack,"",@progbits
+```
+
+WARNING: The name is going to be different on 32-bit and 64-bit platforms due to different sizes
+of `size_t`. Also ```void* operator new[](std::size_t)``` is a separate function as well as `malloc()`.
+The latter is a C function and hence not mangled by the way.
+
+`stackcount` doesn't count how much memory is allocated, but how often. To answer the "how much"
+question you could use Brendan's
+[mallocstacks](https://github.com/brendangregg/BPF-tools/blob/master/old/2017-12-23/mallocstacks.py)
+tool, but it works only for `malloc()` calls. You need to modify it to take into
+account other memory allocating functions.
+
+# Performance annotations
+
+In case there is a need to measure how long a code path takes time to execute in Envoy you may
+resort to instrumenting the code with the
+[performance annotations](https://github.com/envoyproxy/envoy/blob/master/source/common/common/perf_annotation.h).
+
+There are two types of the annotations. The first one is used to measure operations limited by
+a common lexical scope. For example:
+
+```c++
+void doHeavyLifting() {
+  PERF_OPERATION(op);
+  bool success = doSomething();
+  if (success) {
+    finalizeSuccessfulOperation();
+    PERF_RECORD(op, "heavy lifting", "successful");
+  } else {
+    recoverFromFailure();
+    PERF_RECORD(op, "heavy lifting", "failed")
+  }
+}
+```
+
+The recorded performance data can be dumped to stdout with a call to `PERF_DUMP()`:
+```
+Duration(us)  # Calls  Mean(ns)  StdDev(ns)  Min(ns)  Max(ns)  Category       Description
+        2837       22    128965     37035.5   109731   241957  heavy lifting  successful
+         204       13     15745      2449.4    13323    20446  heavy lifting  failed
+```
+
+The second type is performance annotations owned by a class instance. They can measure
+operations spanned across the instance's methods:
+```c++
+class CustomFilter : public Http::StreamEncoderFilter {
+public:
+
+  ...
+  // Http::StreamEncoderFilter
+  Http::FilterHeadersStatus encodeHeaders(Http::ResponseHeaderMap& headers,
+                                          bool end_stream) override {
+    PERF_OWNED_OPERATION(operation);
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  Http::FilterDataStatus encodeData(Buffer::Instance& buffer, bool end_stream) override {
+    if (end_stream) {
+      PERF_OWNED_RECORD(operation, "stream encoding", "without trailers")
```

Comment:
consider calling this member-var `perf_operation_`

Reply:
Yeap, renamed.